client/gui/InterfaceObjectConfigurable.cpp

```diff
@@ -0,0 +1,386 @@
+	auto itemsTotal = config["itemsTotal"].Integer();
+	auto value = config["selected"].Integer();
+	bool horizontal = config["orientation"].String() == "horizontal";
+	return std::make_shared<CSlider>(position, length, callbacks.at(config["callback"].String()), itemsVisible, itemsTotal, value, horizontal, style);
+}
+
+std::shared_ptr<CAnimImage> InterfaceObjectConfigurable::buildImage(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	int frame = config["frame"].isNull() ? 0 : config["frame"].Integer();
+	return std::make_shared<CAnimImage>(image, frame, group, position.x, position.y);
+}
+
+std::shared_ptr<CFilledTexture> InterfaceObjectConfigurable::buildTexture(const JsonNode & config) const
+{
+	auto image = config["image"].String();
+	auto rect = readRect(config["rect"]);
+	return std::make_shared<CFilledTexture>(image, rect);
+}
+
+std::shared_ptr<CShowableAnim> InterfaceObjectConfigurable::buildAnimation(const JsonNode & config) const
+{
+	auto position = readPosition(config["position"]);
+	auto image = config["image"].String();
+	ui8 flags = 0;
+	if(!config["repeat"].Bool())
+		flags |= CShowableAnim::EFlags::PLAY_ONCE;
+	
+	int group = config["group"].isNull() ? 0 : config["group"].Integer();
+	auto anim = std::make_shared<CShowableAnim>(position.x, position.y, image, flags, 4, group);
+	if(!config["alpha"].isNull())
+		anim->setAlpha(config["alpha"].Integer());
+	if(!config["callback"].isNull())
+		anim->callback = std::bind(callbacks.at(config["callback"].String()), 0);
+	if(!config["frames"].isNull())
+	{
+		auto b = config["frames"]["start"].Integer();
+		auto e = config["frames"]["end"].Integer();
+		anim->set(group, b, e);
+	}
+	return anim;
+}
+
+std::shared_ptr<CIntObject> InterfaceObjectConfigurable::buildWidget(JsonNode config) const
+{
+	assert(!config.isNull());
+	//overrides from variables
+	for(auto & item : config["overrides"].Struct())
+	{
+		config[item.first] = variables[item.second.String()];
+	}
+	
+	auto type = config["type"].String();
+	if(type == "picture")
+	{
+		return buildPicture(config);
+	}
+	if(type == "image")
+	{
+		return buildImage(config);
+	}
+	if(type == "texture")
+	{
+		return buildTexture(config);
+	}
+	if(type == "animation")
+	{
+		return buildAnimation(config);
+	}
+	if(type == "label")
+	{
+		return buildLabel(config);
+	}
+	if(type == "toggleGroup")
+	{
+		return buildToggleGroup(config);
+	}
+	if(type == "toggleButton")
+	{
+		return buildToggleButton(config);
+	}
+	if(type == "button")
+	{
+		return buildButton(config);
+	}
+	if(type == "labelGroup")
+	{
+		return buildLabelGroup(config);
+	}
+	if(type == "slider")
+	{
+		return buildSlider(config);
+	}
+	if(type == "custom")
+	{
+		return const_cast<InterfaceObjectConfigurable*>(this)->buildCustomWidget(config);
+	}
+	return std::shared_ptr<CIntObject>(nullptr);
```

Comment:
```suggestion
	logGlobal->error("Failed to construct object of type '%s'", type);
	return std::shared_ptr<CIntObject>(nullptr);
```
What if there is a typo in config file?
No warning, no assert, no throw, just silently create dummy object.

Reply:
Good point, I need to cover this functionality with logging